sbtc-core/src/operations/commit_reveal/construction.rs

```diff
@@ -0,0 +1,151 @@
+/*!
+Construction of commit reveal transactions
+*/
+use std::io;
+
+use bdk::bitcoin::{
+    secp256k1::ecdsa::RecoverableSignature, Address as BitcoinAddress, Amount, Transaction, TxOut,
+    XOnlyPublicKey,
+};
+use stacks_core::{codec::Codec, utils::PrincipalData};
+
+use crate::operations::{
+    commit_reveal::utils::{commit, reveal, CommitRevealResult, RevealInputs},
+    Opcode,
+};
+
+/// Data to construct a commit reveal deposit transaction
+pub struct DepositData {
+    /// Address or contract to deposit to
+    pub principal: PrincipalData,
+    /// How much to send for the reveal fee
+    pub reveal_fee: Amount,
+}
+
+impl Codec for DepositData {
+    fn codec_serialize<W: io::Write>(&self, dest: &mut W) -> io::Result<()> {
+        Codec::codec_serialize(&Opcode::Deposit, dest)?;
+        self.principal.codec_serialize(dest)?;
+        self.reveal_fee.codec_serialize(dest)?;
+
+        todo!()
+    }
+
+    fn codec_deserialize<R: io::Read>(data: &mut R) -> io::Result<Self>
+    where
+        Self: Sized,
+    {
+        let opcode = Opcode::codec_deserialize(data)
+            .map_err(|err| io::Error::new(io::ErrorKind::InvalidData, err))?;
+
+        if !matches!(opcode, Opcode::Deposit) {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "Invalid opcode, expected deposit",
+            ));
+        }
+
+        let principal = PrincipalData::codec_deserialize(data)?;
+        let reveal_fee = Amount::codec_deserialize(data)?;
+
+        Ok(Self {
+            principal,
+            reveal_fee,
+        })
+    }
+}
+
+/// Data to construct a commit reveal withdrawal transaction
+pub struct WithdrawalData {
+    /// Amount to withdraw
+    pub amount: Amount,
+    /// Signature of the transaction
+    pub signature: RecoverableSignature,
+    /// How much to send for the reveal fee
+    pub reveal_fee: Amount,
+}
+
+impl Codec for WithdrawalData {
+    fn codec_serialize<W: io::Write>(&self, dest: &mut W) -> io::Result<()> {
+        self.amount.codec_serialize(dest)?;
+        self.signature.codec_serialize(dest)?;
+        self.reveal_fee.codec_serialize(dest)
+    }
+
+    fn codec_deserialize<R: io::Read>(data: &mut R) -> io::Result<Self>
+    where
+        Self: Sized,
+    {
+        let amount = Amount::codec_deserialize(data)?;
+        let signature = RecoverableSignature::codec_deserialize(data)?;
+        let reveal_fee = Amount::codec_deserialize(data)?;
+
+        Ok(Self {
+            amount,
+            signature,
+            reveal_fee,
+        })
+    }
+}
+
+/// Constructs a peg in payment address
+pub fn deposit_commit(
```

Comment:
I do find it a bit confusing that deposit_commit() doesn't create a deposit_commit transaction. Maybe rename to deposit_commit_address()?